src/compositemark/boxplot.ts

```diff
@@ -100,119 +99,134 @@ export function normalizeBoxPlot(spec: GenericUnitSpec<Encoding<string>, BoxPlot
+        },
+        [continuousAxis + '2']: {
+          field: 'upper_box_' + continuousAxisChannelDef.field,
+          type: continuousAxisChannelDef.type
+        },
+        ...encodingWithoutContinuousAxis,
+        ...(size ? {size} : {})
+      }
+    }, { // median tick
+      mark: {
+        type: 'tick',
+        ...(sizeValue ? {size: sizeValue} : {}),
+        ...getMarkDefMixins<BoxPlotPartsMixins>(markDef, 'median', config.boxplot)
+      },
+      encoding: {
+        [continuousAxis]: {
+          field: 'mid_box_' + continuousAxisChannelDef.field,
+          type: continuousAxisChannelDef.type
+        },
+        ...encodingWithoutSizeColorAndContinuousAxis,
+        ...(size ? {size} : {})
+      }
+    }
+  ];
 
   if (isMinMax) {
-    return result;
+    return {
+      ...outerSpec,
+      transform,
+      layer: boxLayer
+    };
   }
 
+  const lowerBoxStr: string = 'datum.lower_box_' + continuousAxisChannelDef.field;
+  const upperBoxStr: string = 'datum.upper_box_' + continuousAxisChannelDef.field;
+  const iqrStr = `(${upperBoxStr} - ${lowerBoxStr})`;
+  const lowerWhiskerStr = `${lowerBoxStr} - ${extent} * ${iqrStr}`;
+  const upperWhiskerStr = `${upperBoxStr} + ${extent} * ${iqrStr}`;
+
   // tukey plot
   return {
     ...outerSpec,
     layer: [
       { // boxplot
         transform,
-        layer: result.layer
+        layer: boxLayer
       }, {
         transform: [
           {
-            window: transform[0].aggregate.slice(0, 2),
+            window: boxParamsQuartiles(continuousAxisChannelDef.field),
             frame: [null, null],
-            groupby: transform[0].groupby
+            groupby: getGroupbyBoxParams(transform)
           }, {
-            filter: `(datum.${continuousAxisChannelDef.field} < datum.lower_box_${continuousAxisChannelDef.field} - ${extent} * (datum.upper_box_${continuousAxisChannelDef.field} - datum.lower_box_${continuousAxisChannelDef.field})) || (datum.${continuousAxisChannelDef.field} > datum.upper_box_${continuousAxisChannelDef.field} + ${extent} * (datum.upper_box_${continuousAxisChannelDef.field} - datum.lower_box_${continuousAxisChannelDef.field}))`
+            filter: `(datum.${continuousAxisChannelDef.field} < ${lowerWhiskerStr}) || (datum.${continuousAxisChannelDef.field} > ${upperWhiskerStr})`
           }
         ],
         mark: {
           type: 'point',
-          ...(sizeValue ? {size: sizeValue} : {}),
           ...getMarkDefMixins<BoxPlotPartsMixins>(markDef, 'outliers', config.boxplot)
         },
         encoding: {
           [continuousAxis]: {
             field: continuousAxisChannelDef.field,
             type: continuousAxisChannelDef.type
           },
-          ...encodingWithoutSizeColorAndContinuousAxis,
-          ...(size ? {size} : {})
+          ...encodingWithoutSizeColorAndContinuousAxis
         }
       }
     ]
   };
 }
 
+function getGroupbyBoxParams(transform: Transform[]): string[] {
```

Comment:
This O(n) search is quite inefficient and hard to understand why it is correct.  

Better extract a method for boxAggregate.